### PR TITLE
[msal-common]: Add metadata resiliency

### LINKED ISSUE #4686: Request Deduplication
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Wrapper Library

Not Applicable

### Description

Hi!

I have a React project which uses `@azure/msal-browser`. When we call our API, we use a custom fetcher that calls `acquireTokenSilent` before the request. We use the custom fetcher through a library (`SWR`) that includes fetching hooks with request deduplication etc.

So if I use the fetching hook from the library multiple places to fetch data from the same endpoint, our custom fetcher will only be called once. However, if the hook is used multiple times in the same page, to different endpoints, our custom fetcher gets called multiple times, and therefore also `acquireTokenSilent`. This results in multiple network calls for configurations and tokens, and often so many that the browser limits other requests.

(I know i can skip the metadata retrieval by manually providing it, but it still happens for tokens)

image
image
image

I might be implementing the whole thing wrong, and should decouple the token request from the api request in my code. But it would be nice if there was some kind of request deduplication functionality in the package.

### Source

External (Customer)

### PR DESCRIPTION
**Pull request by samuelkubai**
The proposed changes are to add resiliency to how we fetch authority metadata for MSAL.
We are adding a hardcoded list of metadata for common authorities eliminating the need to go to the network to fetch metadata for this common authorities.

### CODE REVIEW

**Review by peterzenz — APPROVED** [review]

File: change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "feat: adding the metadata npm commands",
+  "packageName": "@azure/msal-node-extensions",
+  "email": "kamausamuel11@gmail.com",
```
Comment (jasonnutter) [inline change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json]: Did you mean to use your @microsoft.com email?

File: change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "feat: adding the metadata npm commands",
```
Comment (jasonnutter) [inline change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json]: For all of these: 

```suggestion
  "comment": "feat: adding the metadata npm commands #4536 ",
```

File: lib/msal-browser/src/config/Configuration.ts [hunk 3]
```diff
@@ -63,6 +63,10 @@ export type BrowserAuthOptions = {
      * Enum that represents the Azure Cloud to use.
      */
     azureCloudOptions?: AzureCloudOptions;
+    /**
+     * Flag of whether to use the local metadata cache
+     */
+    skipLocalMetadataCache?: boolean;
```
Comment (jasonnutter) [inline lib/msal-browser/src/config/Configuration.ts]: This is an API option that should probably be consistent across SDKs (and thus have a design review for).

Comment (jasonnutter) [reply lib/msal-browser/src/config/Configuration.ts]: And I would say the name should be more specific, e.g. `skipAuthorityMetadataCache`.

File: lib/msal-common/package.json [hunk 4]
```diff
@@ -68,7 +71,7 @@
     "ts-jest": "^27.0.2",
     "tslib": "^1.10.0",
     "tslint": "^5.20.0",
-    "typescript": "^3.7.5"
-  },
-  "dependencies": {}
+    "typescript": "^3.7.5",
+    "yargs": "^17.5.1"
```
Comment (jasonnutter) [inline lib/msal-common/package.json:75]: Please make this a dev dependency, and move the relevant script outside of the `src` directory, since it is not needed at runtime.

File: lib/msal-common/src/utils/Constants.ts [hunk 5]
```diff
@@ -394,6 +396,63 @@ export enum CacheOutcome {
     REFRESH_CACHED_ACCESS_TOKEN = "4"
 }
 
+export const EndpointMetadataMap = {
+    "https://login.microsoftonline.com/common/" : {
```
Comment (jasonnutter) [inline lib/msal-common/src/utils/Constants.ts]: We should also have `/consumers/` and `/organizations/` in this map.

File: lib/msal-common/src/scripts/metadata.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
```
Comment (jasonnutter) [inline lib/msal-common/src/scripts/metadata.js]: Please move outside of `src` folder.

File: change/@azure-msal-angular-e959400e-9863-454e-ad9b-aaf61716280e.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "feat: adding the metadata npm commands",
```
Comment (tnorling) [inline change/@azure-msal-angular-e959400e-9863-454e-ad9b-aaf61716280e.json]: Angular doesn't appear to have any changes. Let's remove the package-lock changes if they're not relevant to this PR

File: change/@azure-msal-node-c75d4701-e734-4cf0-8cad-7953484e6703.json [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
```
Comment (tnorling) [inline change/@azure-msal-node-c75d4701-e734-4cf0-8cad-7953484e6703.json]: Looks like there's changes to node, please update to the relevant change type

File: change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json (hunk 2 above)
Comment (tnorling) [inline change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json]: No changes to extensions, please remove

File: change/@azure-node-token-validation-92b19c16-3914-44c1-b482-3478f566a425.json [hunk 9]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "feat: adding the metadata npm commands",
```
Comment (tnorling) [inline change/@azure-node-token-validation-92b19c16-3914-44c1-b482-3478f566a425.json]: No changes, please remove

File: change/msal-9dc2773a-eef8-4657-aa20-b98b34f8449f.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "feat: adding the metadata npm commands",
```
Comment (tnorling) [inline change/msal-9dc2773a-eef8-4657-aa20-b98b34f8449f.json]: Please remove package-lock changes and remove this file

File: lib/msal-browser/src/network/FetchClient.ts [hunk 11]
```diff
@@ -20,10 +20,26 @@ export class FetchClient implements INetworkModule {
...
         let response;
+        let controller: AbortController | undefined;
+        let timeoutId: number | undefined;
+
+        /*
+         * Check if a timeout value is provided via the options and create 
+         * a timeout based on the value provided to abort the request
+         */
+        if (options?.timeout) {
+            controller = new AbortController();
```
Comment (tnorling) [inline lib/msal-browser/src/network/FetchClient.ts]: This is not supported in IE11

File: lib/msal-browser/src/network/FetchClient.ts [hunk 12]
```diff
@@ -20,10 +20,26 @@ export class FetchClient implements INetworkModule {
...
+            timeoutId = window.setTimeout(() => controller?.abort(), options.timeout);
```
Comment (jasonnutter) [inline lib/msal-browser/src/network/FetchClient.ts]: If we want to add timeout logic here (which we should), it should be in both GET and POST, and for both the fetch client and xhr client. Given that, it might be better to have this put into a separate PR, with this PR just containing the caching logic.

File: lib/msal-common/src/authority/Authority.ts [hunk 13]
```diff
@@ -353,6 +355,49 @@ export class Authority {
         }
     }
 
+    /**
+     * Get OAuth endpoints for common authorities.
+     */
+    private async getEndpointMetadataFromHardcodedValues(): Promise<OpenIdConfigResponse | null> {
```
Comment (tnorling) [inline lib/msal-common/src/authority/Authority.ts]: Why is this Async?

**Review by jasonnutter — CHANGES_REQUESTED** [review]
Feedback posted. Would also like to see some documentation for this, given it is a change in behavior and introduces new API options.

File: lib/msal-common/src/authority/Authority.ts [hunk 14]
```diff
@@ -408,35 +459,66 @@ export class Authority {
...
+                this.hostnameAndPort
+            );
         }
         return match;
     }
 
+    /**
+     * Get cloud discovery metadata for common authorities 
+     */
+    private async getCloudDiscoveryMetadataFromHarcodedValues(): Promise<CloudDiscoveryMetadata | null> {
```
Comment (tnorling) [inline lib/msal-common/src/authority/Authority.ts]: Does this need to be Async?

File: lib/msal-common/src/scripts/metadata.js [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+const axios = require("axios");
```
Comment (tnorling) [inline lib/msal-common/src/scripts/metadata.js]: Can we move this to `lib/msal-common/scripts`?

Comment (tnorling) [inline lib/msal-common/src/scripts/metadata.js]: Axios has had a string of vulnerabilities lately, any alternatives we can use instead?

File: lib/msal-common/src/utils/Metadata.ts [hunk 16]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import metadataJson from "./metadata.json";
```
Comment (tnorling) [inline lib/msal-common/src/utils/Metadata.ts]: Importing from .json files has been problematic in the past. Can we save the data to a .ts file instead?

File: lib/msal-common/src/utils/Constants.ts [hunk 17]
```diff
@@ -394,6 +396,63 @@ export enum CacheOutcome {
     REFRESH_CACHED_ACCESS_TOKEN = "4"
 }
 
+export const EndpointMetadataMap = {
```
Comment (tnorling) [inline lib/msal-common/src/utils/Constants.ts]: Is this duplicating the content in the metadata.json file? Can we use one or the other?

File: lib/msal-common/src/utils/Constants.ts [hunk 18]
```diff
@@ -394,6 +396,63 @@ export enum CacheOutcome {
...
+    "https://login.microsoftonline.com/common/" : {
+        token_endpoint: "https://login.microsoftonline.com/common/oauth2/v2.0/token",
+        authorization_endpoint: "https://login.microsoftonline.com/common/oauth2/v2.0/authorize",
+        end_session_endpoint: "https://login.microsoftonline.com/common/oauth2/v2.0/logout",
+        issuer: "https://login.microsoftonline.com/{tenantid}/v2.0",
+    }
+};
+
+export const CloudDiscoveryMetadataMap = {
```
Comment (tnorling) [inline lib/msal-common/src/utils/Constants.ts]: Same question here

**Review by tnorling — CHANGES_REQUESTED** [review]

File: .github/workflows/metadata-check.yml [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+# To use this workflow, you will need to set up a .github/labeler.yml
+# file with configuration.  For more information, see:
+# https://github.com/actions/labeler/blob/master/README.md
+
+name: Metadata Checker 
+on:
+  push:
+    branches: 
+      - dev
+      - master
```
Comment (tnorling) [inline .github/workflows/metadata-check.yml]: So that this gets checked on each PR
```suggestion
  pull_request:
    branches: 
      - dev
```

File: .github/workflows/metadata-check.yml [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+concurrency:
+  group: metadata-checker-${{github.ref}}
+  cancel-in-progress: true
+
+jobs:
+  metadata-checker:
+    runs-on: ubuntu-latest
+    if: (github.repository == 'AzureAD/microsoft-authentication-library-for-js')
```
Comment (tnorling) [inline .github/workflows/metadata-check.yml]: This line isn't needed

**Comment by jasonnutter** [comment]
Also, has this been tested with instance_aware?

File: change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json (hunk 1 above)
Comment (samuelkubai) [reply change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json]: I have updated this accordingly

File: change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json (hunk 2 above)
Comment (samuelkubai) [reply change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json]: I have added the ID to all the change files

File: lib/msal-common/package.json (hunk 4 above)
Comment (samuelkubai) [reply lib/msal-common/package.json:75]: It is a dev dependency now.

File: change/@azure-msal-node-c75d4701-e734-4cf0-8cad-7953484e6703.json (hunk 8 above)
Comment (samuelkubai) [reply change/@azure-msal-node-c75d4701-e734-4cf0-8cad-7953484e6703.json]: I have changed it to a "minor" change

File: change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json (hunk 2 above)
Comment (samuelkubai) [reply change/@azure-msal-node-extensions-04dd6516-d72f-43b6-823b-92c7aefd845c.json]: I have removed this change file

File: change/@azure-node-token-validation-92b19c16-3914-44c1-b482-3478f566a425.json (hunk 9 above)
Comment (samuelkubai) [reply change/@azure-node-token-validation-92b19c16-3914-44c1-b482-3478f566a425.json]: I have removed this change file

File: change/msal-9dc2773a-eef8-4657-aa20-b98b34f8449f.json (hunk 10 above)
Comment (samuelkubai) [reply change/msal-9dc2773a-eef8-4657-aa20-b98b34f8449f.json]: I have removed this change file too.

File: .github/workflows/metadata-check.yml (hunk 20 above)
Comment (samuelkubai) [reply .github/workflows/metadata-check.yml]: I have removed this line from the workflow file

File: lib/msal-common/src/utils/Constants.ts (hunk 17 above)
Comment (samuelkubai) [reply lib/msal-common/src/utils/Constants.ts]: You are right, we should use the data in the metadata.json file.

File: lib/msal-common/src/utils/Constants.ts (hunk 18 above)
Comment (samuelkubai) [reply lib/msal-common/src/utils/Constants.ts]: Same here, I have made that change.

File: lib/msal-common/src/scripts/metadata.js (hunk 6 above)
Comment (samuelkubai) [reply lib/msal-common/src/scripts/metadata.js]: I have moved it to the root of `msal-common`

File: lib/msal-common/src/utils/Constants.ts (hunk 5 above)
Comment (samuelkubai) [reply lib/msal-common/src/utils/Constants.ts]: I have added the `/consumers/` and `/organizations/` in the map

File: lib/msal-common/src/scripts/metadata.js (hunk 15 above)
Comment (samuelkubai) [reply lib/msal-common/src/scripts/metadata.js]: I have moved this scripts to the proposed location.

File: lib/msal-common/src/utils/Metadata.ts (hunk 16 above)
Comment (samuelkubai) [reply lib/msal-common/src/utils/Metadata.ts]: We are generating this `json` file directly from the response, this would be difficult to make into a .ts file
What problems have we faced in the past?

File: lib/msal-browser/src/config/Configuration.ts (hunk 3 above)
Comment (samuelkubai) [reply lib/msal-browser/src/config/Configuration.ts]: We do have a design document for this here in this [document](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/MSALJS_V2/Metadata-Resiliency.md&version=GBmetadata-resiliency&_a=preview&anchor=opting-out)

Comment (samuelkubai) [reply lib/msal-browser/src/config/Configuration.ts]: Do we need another document?

File: lib/msal-browser/src/network/FetchClient.ts (hunk 12 above)
Comment (samuelkubai) [reply lib/msal-browser/src/network/FetchClient.ts]: Got it, I have removed that logic for now.

File: change/@azure-msal-angular-e959400e-9863-454e-ad9b-aaf61716280e.json (hunk 7 above)
Comment (samuelkubai) [reply change/@azure-msal-angular-e959400e-9863-454e-ad9b-aaf61716280e.json]: I have removed the package-lock.json file and the associated change file

File: lib/msal-common/src/utils/Metadata.ts (hunk 16 above)
Comment (samuelkubai) [reply lib/msal-common/src/utils/Metadata.ts]: I have added a typescript file for the raw metadata, I believe I encountered the problem you mentioned on building the project with a JSON file.

**Comment by samuelkubai** [comment]
> Also, has this been tested with instance_aware?

@jasonnutter I don't believe I have, could you elaborate on what you mean by `instance_aware`?

> Feedback posted. Would also like to see some documentation for this, given it is a change in behavior and introduces new API options.

And by documentation do you mean like this [here](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/MSALJS_V2/Metadata-Resiliency.md&version=GBmetadata-resiliency&_a=preview&anchor=opting-out) or in the actual codebase?

File: lib/msal-browser/src/config/Configuration.ts (hunk 3 above)
Comment (samuelkubai) [reply lib/msal-browser/src/config/Configuration.ts]: I have renamed the flag accordingly.

File: lib/msal-common/src/scripts/metadata.js (hunk 15 above)
Comment (samuelkubai) [reply lib/msal-common/src/scripts/metadata.js]: I have replaced `axios` here with `https`

File: lib/msal-common/package.json [hunk 21]
```diff
@@ -57,9 +58,11 @@
     "@types/debug": "^4.1.5",
     "@types/jest": "^26.0.23",
     "@types/sinon": "^7.5.0",
+    "axios": "^0.27.2",
```
Comment (tnorling) [inline lib/msal-common/package.json]: This can be uninstalled since you moved to https

File: lib/msal-common/src/utils/metadata.json [hunk 22]
```diff
@@ -0,0 +1 @@
...rosoftonline.us","login.usgovcloudapi.net"]},{"preferred_network":"login-us.microsoftonline.com","preferred_cache":"login-us.microsoftonline.com","aliases":["login-us.microsoftonline.com"]}]},"https://login.microsoftonline.us/organizations/":{"tenant_discovery_endpoint":"https://login.microsoftonline.us/organizations/v2.0/.well-known/openid-configuration","api-version":"1.1","metadata":[{"preferred_network":"login.microsoftonline.com","preferred_cache":"login.windows.net","aliases":["login.microsoftonline.com","login.windows.net","login.microsoft.com","sts.windows.net"]},{"preferred_network":"login.partner.microsoftonline.cn","preferred_cache":"login.partner.microsoftonline.cn","aliases":["login.partner.microsoftonline.cn","login.chinacloudapi.cn"]},{"preferred_network":"login.microsoftonline.de","preferred_cache":"login.microsoftonline.de","aliases":["login.microsoftonline.de"]},{"preferred_network":"login.microsoftonline.us","preferred_cache":"login.microsoftonline.us","aliases":["login.microsoftonline.us","login.usgovcloudapi.net"]},{"preferred_network":"login-us.microsoftonline.com","preferred_cache":"login-us.microsoftonline.com","aliases":["login-us.microsoftonline.com"]}]}}}
```
Comment (tnorling) [inline lib/msal-common/src/utils/metadata.json]: Please remove this file

File: lib/msal-common/src/utils/Metadata.ts [hunk 23]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { rawMetdataJSON } from "./RawMetadata";
```
Comment (tnorling) [inline lib/msal-common/src/utils/Metadata.ts]: Can we combine this file with `RawMetadata.ts`, rename to `AuthorityMetadata.ts` and move to the `Authority` folder?

**Review by tnorling — COMMENTED** [review]
A couple more suggestions, otherwise looks good. Thanks for addressing feedback!

File: lib/msal-common/package.json (hunk 21 above)
Comment (samuelkubai) [reply lib/msal-common/package.json]: I have removed the unused dependency from the `package.json` file

File: lib/msal-common/src/utils/Metadata.ts (hunk 23 above)
Comment (samuelkubai) [reply lib/msal-common/src/utils/Metadata.ts]: Good idea, I have merged the two files `Metadata.ts` and `RawMetadata.ts` files into `AuthorityMetadata.ts`

File: lib/msal-common/src/utils/metadata.json (hunk 22 above)
Comment (samuelkubai) [reply lib/msal-common/src/utils/metadata.json]: Having this discussion over teams, to find the best way forward.

Comment (samuelkubai) [reply lib/msal-common/src/utils/metadata.json]: I have found a solution on how to merge the single sources of truth.

File: lib/msal-browser/src/network/FetchClient.ts (hunk 11 above)
Comment (samuelkubai) [reply lib/msal-browser/src/network/FetchClient.ts]: This is no longer part of this pull request now, we will address it then.

File: lib/msal-common/src/authority/Authority.ts (hunk 13 above)
Comment (samuelkubai) [reply lib/msal-common/src/authority/Authority.ts]: Right, the `async` here is unnecessary.

File: lib/msal-common/src/authority/Authority.ts (hunk 14 above)
Comment (samuelkubai) [reply lib/msal-common/src/authority/Authority.ts]: Right, `async` here is not required.

**Review by tnorling — APPROVED** [review]
Looks good! Thanks for addressing feedback

**Comment by jasonnutter** [comment]
> > Also, has this been tested with instance_aware?
> 
> @jasonnutter I don't believe I have, could you elaborate on what you mean by `instance_aware`?
> 
> > Feedback posted. Would also like to see some documentation for this, given it is a change in behavior and introduces new API options.
> 
> And by documentation do you mean like this [here](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/MSALJS_V2/Metadata-Resiliency.md&version=GBmetadata-resiliency&_a=preview&anchor=opting-out) or in the actual codebase?

Instance aware is the feature that allows an application to work in both public and sovereign clouds, by telling ests to tell MSAL which cloud the user authenticated to. This would mean that the authority used to acquire tokens could be different than the one configured. I would like to see tests that ensure that flow isn't broken: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/instance-aware.md

And was referring to user facing documation, namely this file: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/configuration.md#configuration-options-1

File: lib/msal-common/src/authority/Authority.ts [hunk 24]
```diff
@@ -272,40 +273,32 @@ export class Authority {
...
-                    metadata = Authority.replaceWithRegionalInformation(metadata, azureRegion);
-                }
+                metadata = await this.updateMetadataWithRegionalInformation(metadata);
             }
 
             metadataEntity.updateEndpointMetadata(metadata, true);
             return AuthorityMetadataSource.NETWORK;
+        }    
+
+        if (harcodedMetadata && !this.authorityOptions.skipAuthorityMetadataCache) {
```
Comment (jasonnutter) [inline lib/msal-common/src/authority/Authority.ts:300]: I don't see new unit tests that test this code path. Please add tests showing if this code path is hit, the hardcoded meta data is used.

Comment (jasonnutter) [reply lib/msal-common/src/authority/Authority.ts:300]: This is where `instance_aware` should probably be checked as well. I would think that if instance_aware is enabled, we should not use the hardcoded metadata.

File: lib/msal-browser/src/app/IPublicClientApplication.ts [hunk 25]
```diff
@@ -99,6 +101,7 @@ export const stubbedPublicClientApplication: IPublicClientApplication = {
     ssoSilent: () => {	
         return Promise.reject(BrowserConfigurationAuthError.createStubPcaInstanceCalledError());	
     },
+    // eslint-disable-next-line react/no-multi-comp
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/IPublicClientApplication.ts]: Strange. We should figure out whats going on here, this should defintely not be needed.

Comment (tnorling) [reply lib/msal-browser/src/app/IPublicClientApplication.ts]: The offending library has been downgraded to address. Please remove the ignores from this PR. Thanks!

File: lib/msal-browser/src/config/Configuration.ts (hunk 3 above)
Comment (bktowett) [reply lib/msal-browser/src/config/Configuration.ts]: @jasonnutter was this resolved?

Comment (jasonnutter) [reply lib/msal-browser/src/config/Configuration.ts]: The new name is good. Not sure if it has been agreed upon for other SDKs (or if that is something we should be concerned about right now). @peterzenz thoughts?

File: lib/msal-browser/src/app/IPublicClientApplication.ts (hunk 25 above)
Comment (jo-arroyo) [reply lib/msal-browser/src/app/IPublicClientApplication.ts]: Ignores removed. Thanks!

**Review by jasonnutter — APPROVED** [review]

**Comment by jo-arroyo** [comment]
Documentation for this feature will be added incrementally.

File: lib/msal-common/src/authority/Authority.ts (hunk 24 above)
Comment (jo-arroyo) [reply lib/msal-common/src/authority/Authority.ts:300]: Unit tests for new code paths added. 

Offline discussion with Jason concluded that separate `instance_aware` check was not needed here, as `cloud_instance_host_name` host name on response results in updating metadata.
